cmd/aro/update_role_sets.go

```diff
@@ -0,0 +1,174 @@
+	serviceKeyvault := keyvault.NewManager(msiKVAuthorizer, serviceKeyvaultURI)
+
+	aead, err := encryption.NewMulti(ctx, serviceKeyvault, env.EncryptionSecretV2Name, env.EncryptionSecretName)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := env.ValidateVars(envDatabaseAccountName); err != nil {
+		return nil, err
+	}
+	dbAccountName := os.Getenv(envDatabaseAccountName)
+	clientOptions := &policy.ClientOptions{
+		ClientOptions: _env.Environment().ManagedIdentityCredentialOptions().ClientOptions,
+	}
+
+	logrusEntry := log.WithField("component", "database")
+	dbAuthorizer, err := database.NewMasterKeyAuthorizer(ctx, logrusEntry, msiToken, clientOptions, _env.SubscriptionID(), _env.ResourceGroup(), dbAccountName)
+	if err != nil {
+		return nil, err
+	}
+
+	dbc, err := database.NewDatabaseClient(log.WithField("component", "database"), _env, dbAuthorizer, m, aead, dbAccountName)
+	if err != nil {
+		return nil, err
+	}
+
+	dbName, err := DBName(_env.IsLocalDevelopmentMode())
+	if err != nil {
+		return nil, err
+	}
+
+	return database.NewPlatformWorkloadIdentityRoleSets(ctx, dbc, dbName)
+}
+
+func updatePlatformWorkloadIdentityRoleSetsInCosmosDB(ctx context.Context, dbPlatformWorkloadIdentityRoleSets database.PlatformWorkloadIdentityRoleSets, log *logrus.Entry) error {
+	// 2 - Get the existing role set documents, if existing
+	// Mostly copied from update_ocp_versions.go
+	existingRoleSets, err := dbPlatformWorkloadIdentityRoleSets.ListAll(ctx)
+	if err != nil {
+		return nil
+	}
+
+	incomingRoleSets, err := getRoleSetsFromEnv()
+	if err != nil {
+		return err
+	}
+
+	newRoleSets := make(map[string]api.PlatformWorkloadIdentityRoleSetProperties)
+	for _, doc := range incomingRoleSets {
+		newRoleSets[doc.OpenShiftVersion] = doc
+	}
+
+	// 3 - Put/patch the new role sets to the doc, overwriting whatever is there for that version, or adding if new
+	// Mostly copied from update_ocp_versions.go
+	for _, doc := range existingRoleSets.PlatformWorkloadIdentityRoleSetDocuments {
+		existing, found := newRoleSets[doc.PlatformWorkloadIdentityRoleSet.Properties.OpenShiftVersion]
+		if found {
+			log.Printf("Found Version %q, patching", existing.OpenShiftVersion)
+			_, err := dbPlatformWorkloadIdentityRoleSets.Patch(ctx, doc.ID, func(inFlightDoc *api.PlatformWorkloadIdentityRoleSetDocument) error {
+				inFlightDoc.PlatformWorkloadIdentityRoleSet.Properties = existing
+				return nil
+			})
+			if err != nil {
+				return err
+			}
+			log.Printf("Version %q found", existing.OpenShiftVersion)
+			delete(newRoleSets, existing.OpenShiftVersion)
+			continue
+		}
+
+		log.Printf("Version %q not found, deleting", doc.PlatformWorkloadIdentityRoleSet.Properties.OpenShiftVersion)
+		// Delete via changefeed
+		_, err := dbPlatformWorkloadIdentityRoleSets.Patch(ctx, doc.ID,
+			func(d *api.PlatformWorkloadIdentityRoleSetDocument) error {
+				d.PlatformWorkloadIdentityRoleSet.Deleting = true
+				d.TTL = 60
+				return nil
+			})
```

Comment:
It's not clear to me why it's being deleted via a Patch rather than using the actual delete method. Can you flesh this out for me?

Reply:
Sure - we don't ever delete docs from CosmosDB directly, but instead patch them with Deleting and a TTL where it is then cleaned up by CosmosDB itself. We have this convention in the update_ocp_versions script as well.

Reply:
Hmmm. Why the delete method then? From a quick search we never use the Delete method. I wonder would wrapping the "deleting patch" in the Delete method be clearer? Doesn't block this at all but I've made a note for myself